### PR TITLE
[WIP] CP4MCM 11519

### PR DESCRIPTION
**Pull request by morningspace**
https://github.ibm.com/IBMPrivateCloud/CP4MCM/issues/11519

The following script aids the customer to enable HA for CP4MCM. Execute the script with --help to get a rundown of how to use it.

```console
Usage ./cp4mcm-ha-utility.sh
Use this script to enable or diable HA for IBM Cloud Pak for Multicloud Management

  Flags:

     --kubeconfigPath                 The absolute path to the kubeconfig file to access the cluster. Use default kubeconfig if it is omitted
     --list                           List the modules that support HA
     --enable                         Enable HA for all modules, or a particular module if --module is specified
     --disable                        Disable HA for all modules, or a particular module if --module is specified
     --verify                         Verify the results by listing the corresponding Kubernetes resources
     --module <module>                Enable, diable, or verify HA for a particular module. Use --list to get all valid module names
     --help                           Print the help information

  Usage examples:

     ./cp4mcm-ha-utility.sh --list                        List the modules that support HA
     ./cp4mcm-ha-utility.sh --enable                      Enable HA for all modules
     ./cp4mcm-ha-utility.sh --disable                     Disable HA for all modules
     ./cp4mcm-ha-utility.sh --verify                      Verify the results for all modules
     ./cp4mcm-ha-utility.sh --module ui --enable          Enable HA for one module
     ./cp4mcm-ha-utility.sh --module ui --disable         Disable HA for one module
     ./cp4mcm-ha-utility.sh --module ui --verify          Verify the results for one module
```

### CODE REVIEW

File: scripts/cp4mcm-enable-ha.sh [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    return 0
+}
+
+# end admin stuff
+
+# begin core stuff
+
+enableHA(){
+  echo "Start to enable HA for each component in IBM Cloud Pak for Multicloud Management..." | tee -a "$logpath"
+  echo "" | tee -a "$logpath"
```
Comment (gyliu513) [inline scripts/cp4mcm-enable-ha.sh:125]: Can you add some logic here to enable HA to speed up test? 👍

Comment (morningspace) [reply scripts/cp4mcm-enable-ha.sh:125]: Sure, working on it, will update it later today

**Comment by stacyped** [comment]
Is this ready to merge?

**Comment by morningspace** [comment]
Not ready for merge yet. I will add a few more components HA enabling, then make some refactoring, before it get merged. Hopefully to get it done today.

**Comment by gyliu513** [comment]
@stacyped can you help merge this one? We can create follow up PRs later.

**Comment by morningspace** [comment]
@taylormgeorge91 Thanks for your comments. I was standardizing the way to update the component resources into a common method in the script, so that all components should be able to update their resources consistently.

Regarding patching CSV, it appears to me, so far the changes made against CSV are all from the need for each component to scale up its operator, which seems to be a very particular case.

I was thinking to update the corresponding deployment `replicas` value for each operator component, but that wouldn't work, since it will be reconciled based on the deployment `replicas` value defined in CSV. So, I can imagine why they are updating CSV. But yes, a new version update from the channel will replace the patched CSV. How are we gonna address this issue?

**Comment by morningspace** [comment]
Code has been ready for the final review. @gyliu513 

I've added all the components to the script, w/ a bit more extended capabilities for the script. Now, it can be used to:
* List the modules that support HA
* Enable HA for all components or a particular component
* Disable HA for all components or a particular component
* Verify results by listing the corresponding Kubernetes resources

**Comment by gyliu513** [comment]
Looks great, thanks @morningspace 

@taylormgeorge91 for your comments, we can document to ask the customer to run the script again to update the operator.

@stacyped can you help merge this? Thanks.

**Comment by taylormgeorge91** [comment]
@morningspace That makes sense. If the deployment of the operator is updated directly it would be out of sync with the CSV and thus prone to reconciliation. So it makes sense that you would have to patch the CSV and scale up the operator deployment to make the change stick (until next CSV update). 

For CSV update, it would be ideal for us to not revert the changes of replicas. I wonder if there is a way for us to set this in the `subscription` for the operator. If so, we can open an enhancement on ODLM to support subscription configuration at which point we can pass through replica configuration for the operators in this manner. 

Otherwise, we will have to identify another means of persisting the configuration.

File: scripts/cp4mcm-ha-utility.sh [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    if [[ $operate == disable ]]; then
+        echo 1
+    else
+        [[ $1 == mcm ]] && echo 3 || echo 2
+    fi
+}
+
+enableHAForModule() {
+    local namespace="$1"
+    local module="$2"
```
Comment (taylormgeorge91) [inline scripts/cp4mcm-ha-utility.sh:267]: As a note this is the operator/operand and not the module. 

For example, notary is an operator in the security services module.

Comment (morningspace) [reply scripts/cp4mcm-ha-utility.sh:267]: Thanks @taylormgeorge91, I'm removing `Module` from the method name to avoid the possible confusion.

**Comment by morningspace** [comment]
@stacyped Can we get this merged since @gyliu513 has approved, and I don't have to resolve the merge conflict for CHANGELOG.md from time to time :-) Thanks.

**Pull request merged**